cdisc_rules_engine/models/rule_conditions/condition_composite.py

```diff
@@ -24,13 +25,39 @@ class ConditionComposite(ConditionInterface):
     def __init__(self):
         self._conditions = {}
 
+    def copy(self):
+        new_condition = ConditionComposite()
+        for key, condition in self._conditions.items():
+            new_condition.add_conditions(key, condition)
+        return new_condition
+
     def add_conditions(self, key: str, conditions: List[ConditionInterface]):
         """
         Adds a list of objects with ConditionInterface
         interface to the given key.
         """
         self._conditions[key] = conditions
 
+    def add_conditions_for_targets(self, targets: List[str]):
```

Comment:
I moved it here to isolate the logic for determining whether a condition should be duplicated to the individual condition rather than the rule processor needing to determine whether a rule should be duplicated

Reply:
Anyway, I have noticed a couple of things:

1. The idea of composite pattern is to avoid if/else blocks and type checks when working with the tree. Both: nodes and a composite should implement the same interface with the difference that a node executes the operation and a composite executes it for each node. This is why I think we can refactor `add_conditions_for_targets` method in a more elegant way. 
2. Also, I see that composite has `copy` method whereas SingleCondition has `duplicate`. It looks confusing as for me and we can solve this problem using the approach from point 1.
3. Finally, I don't see `add_operator` method neither in interface nor in composite whereas it is being called in `NotConditionComposite`.

Reply:
Ahh all right. I will work on that refactor. In the meantime I will do a release with the latest business rules and development branch so that we can at least unblock date validation

Reply:
@alexfurmenkov I have moved the logic to the rule processor and updated the logic accordingly